# 7️⃣5️⃣ Item 75 : 예외의 상세 메시지에 실패 관련 정보를 담으라

<br>

## 📌 목차
1. 실패 순간과 예외
2. 실패 순간 포착 방법 1
3. 실패 순간 포착 방법 2
4. 실패 순간 포착 방법 3
5. 예외 메시지 vs 오류 메시지

<br>

## 🚨 실패 순간과 예외

예외 터지면 자바 시스템은 예외의 스택 추적 = stack trace 정보를 자동으로 출력한다. 

스택 추적은 예외 객체의 toString 메서드를 호출해 얻는 문자열이다. 

보통은 예외의 클래스 이름 뒤에 상세 메시지가 붙는 형태다. 

모든 예외의 상위 클래스인 Throwable에 가보면 아래와 같이 toString이 있다. 

```java
    public String toString() {
        String s = getClass().getName();
        String message = getLocalizedMessage();
        return (message != null) ? (s + ": " + message) : s;
    }
```

<br>

이 스택 추적 정보가 실패 원인을 분석해야 하는 프로그래머나 SRE = 사이트 신뢰성 엔지니어가 얻을 수 있는 유일한 정보인 경우가 많다. 

실패를 재현할 수 있다면 예외가 발생한 곳 근처로 log 찍어서 문제의 원인을 찾을 수 있다. 

하지만 실패를 재현하기 어렵다면 더 자세한 정보를 얻기 어렵거나 불가능하다. 

그래서 스택 추적 정보가 유일한 정보인 것이다. 

<br>

따라서 예외의 toString 메서드 = 스택 추적에 실패 원인에 관한 정보를 가능한 많이 담는 것이 아주 중요하다. 

사후 분석을 위해 실패 순간의 상황을 정확히 포착해 예외의 상세 메시지 = 스택 추적에 담아야 한다.

<br>

그러면 실패 순간을 어떻게 잘 포착할 수 있는지 알아보자. 

즉 예외가 발생했을 때 어떻게 예외 메시지에 실패 원인 관한 정보 많이 넣을 수 있는지 보자는 것이다. 

<br>

## 1️⃣ 실패 순간 포착 방법 1

실패 순간을 잘 포착할 수 있는 첫번째 방법은 다음과 같다. 

- 발생한 예외에 관여된 모든 매개변수와 필드의 값을 실패 메시지에 담아야 한다.

<br>

예를 들어 IndexOutOfBoundsException이 발생했을 때 실패 메시지에 담겨야 할 것은 아래와 같다. 

1. 범위의 최솟값
2. 범위의 최댓값
3. 범위를 벗어난 index의 값

<br>

이런 정보는 발생한 실패에 관해 많은 것을 알려준다. 

셋 중 한두 개 혹은 셋 모두가 잘못 됐을 수 있다. 

발생한 실패는 모두 원인이 다를 것이다. 

그래서 이런 정보들이 발생한 실패에서 어디를 고쳐야 하는지 분석하는데 큰 도움이 된다. 

<br>

그냥 IndexOutOfBoundsException 발생만 했다라고 하면 3개 중 어디가 문제인지 알 수 없을 것이다. 

그러면 실패 해결하기 위해 다 뒤져봐야 하는 번거로움이 있을 것이다. 

그런데 이런 정보들을 담으면 해결에 큰 도움이 되는 것이다. 

<br>

하지만 실패 메시지에 정보 담을 때 주의할 점이 2가지가 있다. 

1. 보안과 관련한 정보는 주의해서 다뤄야 한다. 
2. 관련 데이터를 담아야 하지만 장황할 필요는 없다. 

<br>

1번 주의점은 보안과 관련한 정보는 주의해서 다뤄야 한다는 것이다. 

문제를 진단하고 해결하는 과정에서 스택 추적 정보는 많은 사람이 볼 수 있다. 

따라서 실패 메시지에 비밀번호나 암호 키 같은 보안과 관련한 정보는 담으면 안된다는 것이다. 

<br>

2번 주의점은 관련 데이터를 모두 담아야 하지만 장황할 필요는 없다는 것이다. 

예외가 발생하면 문제를 분석하는 사람은 아래의 것들을 확인한다. 

1. 스택 추적 정보 (실패 메시지)
2. 관련 문서
3. 소스코드

<br>

그래서 실패 메시지에 문서와 소스코드에서 얻을 수 있는 정보는 넣어봐야 실패 메시지만 길어지게 된다. 

스택 추적 정보에 많은 정보들이 있다. (ex: 예외 발생 줄번호, 호출한 메서드의 줄번호 등…)

그렇기 때문에 이것들을 가지고 관련 문서, 소스코드 보게 하는 것이 낫지 그런 정보들까지 다 담게 되면 불필요한 것만 늘어나는 것이다. 

<br>

## 2️⃣ 실패 순간 포착 방법 2

실패 순간을 잘 포착할 수 있는 두번째 방법은 다음과 같다. 

- 필요한 정보를 예외 생성자에서 모두 받아 상세 메시지까지 미리 생성해 놓는 것이다.

<br>

IndexOutOfBoundsException의 생성자를 보면 아래와 같이 4개가 있다. 

1. 기본 생성자
2. String message를 받는 생성자
3. int Index 값을 받는 생성자
4. long index 값을 받는 생성자. 

<br>

```java
public IndexOutOfBoundsException() { super(); }

public IndexOutOfBoundsException(String s) { super(s); }

public IndexOutOfBoundsException(int index) { super("Index out of range: " + index); }
    
public IndexOutOfBoundsException(long index) { super("Index out of range: " + index); }
```

<br>

하지만 아쉽게도 발생한 예외에 관여될 수 있는 정보인 범위의 최댓값, 최솟값까지 받는 생성자는 없다. 

그래서 아래와 같은 생성자도 있으면 좋다는 것이다. 

```java
public IndexOutOfBoundsException(int lowerBound, int upperBound, int index) { 
	super(String.format(
			"최솟값: %d, 최댓값: %d, 인덱스: %d",
			lowerBound, upperBound, index)); 
}
```

<br>

이렇게 실패 메시지에 담을 모든 정보를 받는 생성자가 있는 것을 강력히 권장한다고 한다. 

이렇게 해두면 프로그래머가 던지는 예외는 자연스럽게 실패를 더 잘 포착한다고 한다. 

<br>

또한 고품질의 상세 메시지를 만들어내는 코드를 예외 클래스 안으로 모아주는 효과도 있다. 

그래서 클래스 사용자가 메시지를 만드는 작업을 중복하지 않아도 된다. 

<br>

## 3️⃣ 실패 순간 포착 방법 3

실패 순간을 잘 포착할 수 있는 세번째 방법은 다음과 같다. 

- 예외는 실패와 관련한 정보를 얻을 수 있는 접근자 메서드를 적절히 제공하는 것이 좋다.

<br>

IndexOutOfBoundsException의 경우에는 lowerBound, upperBound, index를 제공하는 접근자 메서드를 제공하면 좋을 것이다. 

이런 정보들을 실패 메시지에 담는 이유는 예외를 복구하기 위함이다. 

실패 메시지에서 포착한 실패 정보는 예외 상황을 복구하는데 유용할 수 있으므로 접근자 메서드를 제공하는 것이다. 

<br>

접근자 메서드로 받는 정보는 복구에 필요하니 비검사 예외보다는 검사 예외에서 유용할 것이다. 

비검사 예외의 상세 정보에 프로그램적으로 접근하길 원하는 프로그래머는 드물 것이다. 

하지만 ‘toString이 반환하는 값에 포함된 정보 얻어올 수 있는 API를 제공하자’는 Item 12의 조언이 있었다. 

이 조언을 따른다는 관점에서 비검사 예외라도 상세 정보를 알려주는 접근자 메서드를 제공하는 것이 좋다. 

<br>

## 🆚 예외 메시지 vs 오류 메시지

예외의 상세 메시지와 최종 사용자에게 보여줄 오류 메시지를 혼동하면 안된다. 

예외 메시지의 특징 다음과 같다. 

- 가독성보다는 담긴 내용이 훨씬 중요하다.
    - 왜냐하면 예외 메시지는 문제 분석할 프로그래머 등이 보기 때문이다.
- 현지어로 번역하는 일이 거의 없다.

<br>

오류 메시지의 특징은 다음과 같다. 

- 친절한 안내 메시지를 보여줘야 한다.
    - 왜냐하면 오류 메시지는 최종 사용자에게 보여줄 것이기 때문이다.
- 현지어로 번역해주기도 한다.

<br>

예를 들어 IndexOutOfBoundsException가 발생하면 예외 메시지가 아래처럼 나오는 것이다. 

```java
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index 2 out of bounds for length 2
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at Main.main(Main.java:10)

```

<br>

프로그래머는 이 예외 메시지를 보고 이해하고 예외를 처리할 수 있다. 

하지만 최종 사용자에게 문제가 발생했다고 알릴 때 이 예외 메시지를 넘기면 안된다. 

최종 사용자는 저런 예외 메시지를 봐도 이해할 수 없기 때문이다. 

<br>

따라서 최종 사용자에게는 아래처럼 예외 메시지가 아니라 오류 메시지를 보내야 하는 것이다.

```java
"입력할 수 있는 상품의 최대 개수는 2개 입니다."
```